### PR TITLE
Circleci test using new base image

### PR DESCRIPTION
**Pull request by yeounoh**
This PR addresses the following:
- A new ptxla build image `Dockerfile` and `cloudbuild.yaml` for publishing at `gcr.io/tpu-pytorch/xla_base`
- Circleci ci/cd configuration `config.yml` to use the new build image and `gcr` repo.

I've made a few additional changes, such as:
- Change circleci build stage names to be more generic/appropriate
- Change circleci VM definition so that the build runs with a larger VM

The new CI workflow also stages the latest build image in PyTorch image repo in ECR, in case we want to use it in the upstream testing.

### CODE REVIEW

**Comment by JackCaoG** [comment]
@yeounoh seems like absl is missing. A quickly way to test was to comment out test that already passed in `run_test.sh`, so we don't need to rerun them on circle CI. I would expect most of the test failure to be library depended at this point.

**Comment by yeounoh** [comment]
DO_NOT_MERGE_YET, let me check if we can further reduce the image size and reinstate the cache.

**Comment by yeounoh** [comment]
> @yeounoh seems like absl is missing. A quickly way to test was to comment out test that already passed in `run_test.sh`, so we don't need to rerun them on circle CI. I would expect most of the test failure to be library depended at this point.

The GPU tests weren't pi

**Comment by yeounoh** [comment]
> @yeounoh seems like absl is missing. A quickly way to test was to comment out test that already passed in `run_test.sh`, so we don't need to rerun them on circle CI. I would expect most of the test failure to be library depended at this point.

I was actually looking at the GPU tests. The tests weren't detecting, picking up the devices, and all GPU tests were skipping. Now it's resolved, AWS GPU VM has its own guidance/way of surfacing it.

**Comment by yeounoh** [comment]
Hi @seemethere before we merge this PR, is there a way to point a PyTorch PR to point to this, so we can also check if this would pass the PyTorch CI/CD tests?

File: .circleci/config.yml [hunk 1]
```diff
@@ -69,18 +73,25 @@ download_docker: &download_docker
 calculate_docker_image: &calculate_docker_image
   name: Calculate docker image
   command: |
-    git clone --quiet https://github.com/pytorch/pytorch.git "/tmp/pytorch"
     # Just a quick smoke test to see if we can actually extract the tag
+    git clone --quiet https://github.com/pytorch/pytorch.git "/tmp/pytorch"
     git -C /tmp/pytorch rev-parse HEAD:.circleci/docker
     DOCKER_TAG=$(git -C /tmp/pytorch rev-parse HEAD:.circleci/docker)
-    echo "declare -x DOCKER_IMAGE=308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-bionic-cuda10.2-cudnn7-py3.7-clang9:${DOCKER_TAG}" >> "${BASH_ENV}"
+    echo "declare -x ECR_DOCKER_IMAGE=308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/xla_base:${DOCKER_TAG}" >> "${BASH_ENV}"
```
Comment (JackCaoG) [inline .circleci/config.yml:80]: I am a bit confuse about the `ECR_DOCKER_IMAGE ` and `GCR_DOCKER_IMAGE `. Maybe it is worth it to add some comment somewhere in this file?

File: test/cpp/CMakeLists.txt [hunk 2]
```diff
@@ -1,5 +1,9 @@
 cmake_minimum_required(VERSION 3.0)
 
+if( NOT CXX )
+  set(CMAKE_CXX_COMPILER clang++)
```
Comment (JackCaoG) [inline test/cpp/CMakeLists.txt:4]: why is this needed?

**Review by JackCaoG — COMMENTED** [review]
Mostly LGTM

File: .circleci/config.yml [hunk 3]
```diff
@@ -8,23 +8,28 @@ update_submodule: &update_submodule
...
 launch_docker_and_build: &launch_docker_and_build
   name: Launch Docker Container and Build
   no_output_timeout: "1h"
   command: |
     set -e
     eval $(aws ecr get-login --region us-east-1 --no-include-email)
-    time docker pull ${DOCKER_IMAGE} >/dev/null
+    time docker pull ${GCR_DOCKER_IMAGE} >/dev/null
+    time docker tag ${GCR_DOCKER_IMAGE} ${ECR_DOCKER_IMAGE} >/dev/null
+    time docker push ${ECR_DOCKER_IMAGE} >/dev/null
```
[the remark this answers is not in the document]
Comment (yeounoh) [reply .circleci/config.yml:22]: This is to stage our base image for use in the upstream. Right now, we are using their base image in the upstream test -- but, in case we also want the upstream workflow to use our build image, we need to have them in the ECR. This is not expensive, and only pushes it if the image layers are not present in the repo.

File: test/cpp/CMakeLists.txt (hunk 2 above)
Comment (yeounoh) [reply test/cpp/CMakeLists.txt:4]: This is for the upstream test -- they have different compiler flags and settings. e.g., we set `CXX=clang-8` which is not common :)

**Comment by yeounoh** [comment]
Removing DO_NOT_MERGE_YET label.

File: .circleci/common.sh [hunk 4]
```diff
@@ -49,12 +49,16 @@ function checkout_torch_pin_if_available() {
...
   XLA_DIR=$1
+
   # Install ninja to speedup the build
   pip install ninja
 
   # Install libraries required for running some PyTorch test suites
   pip install hypothesis
   pip install cloud-tpu-client
+  pip install absl-py
+  pip install --upgrade numpy>=1.18.5
```
Comment (JackCaoG) [inline .circleci/common.sh:60]: Do you want to rebase and remove my `install_numpy` in this file?

File: .circleci/config.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (JackCaoG) [reply .circleci/config.yml:22]: Can you add this comment to the yml file as comment?

File: .circleci/common.sh (hunk 4 above)
Comment (yeounoh) [reply .circleci/common.sh:60]: Great, will do!

File: .circleci/config.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (JackCaoG) [reply .circleci/config.yml:27]: nit, replace `their` to `pytorch` and `our` to `torch/xla`

**Review by JackCaoG — APPROVED** [review]
Thanks! Let's hope CI passed again

**Review by seemethere — APPROVED** [review]
